packages/contracts/src/contracts/current/protocol/Exchange/MixinExchangeCore.sol

```diff
@@ -126,118 +331,116 @@ contract MixinExchangeCore is
-
-        // Validate order availability
-        uint256 remainingTakerAssetFillAmount = safeSub(order.takerAssetAmount, filled[orderHash]);
-        if (remainingTakerAssetFillAmount == 0) {
-            emit ExchangeError(uint8(Errors.ORDER_FULLY_FILLED), orderHash);
-            return fillResults;
-        }
-
-        // Validate fill order rounding
-        fillResults.takerAssetFilledAmount = min256(takerAssetFillAmount, remainingTakerAssetFillAmount);
-        if (isRoundingError(fillResults.takerAssetFilledAmount, order.takerAssetAmount, order.makerAssetAmount)) {
-            emit ExchangeError(uint8(Errors.ROUNDING_ERROR_TOO_LARGE), orderHash);
-            fillResults.takerAssetFilledAmount = 0;
-            return fillResults;
-        }
-
+    /// @dev Updates state with results of a fill order.
+    /// @param order that was filled.
+    /// @param takerAddress Address of taker who filled the order.
+    /// @param orderTakerAssetFilledAmount Amount of order already filled.
+    /// @return fillResults Amounts filled and fees paid by maker and taker.
+    function updateFilledState(
+        Order memory order,
+        address takerAddress,
+        bytes32 orderHash,
+        uint256 orderTakerAssetFilledAmount,
+        FillResults memory fillResults
+    )
+        internal
+    {
         // Update state
-        filled[orderHash] = safeAdd(filled[orderHash], fillResults.takerAssetFilledAmount);
-
-        // Settle order
-        (fillResults.makerAssetFilledAmount, fillResults.makerFeePaid, fillResults.takerFeePaid) =
-            settleOrder(order, takerAddress, fillResults.takerAssetFilledAmount);
+        filled[orderHash] = safeAdd(orderTakerAssetFilledAmount, fillResults.takerAssetFilledAmount);
 
         // Log order
-        emitFillEvent(order, takerAddress, orderHash, fillResults);
-        return fillResults;
+        emit Fill(
+            order.makerAddress,
+            takerAddress,
+            order.feeRecipientAddress,
+            fillResults.makerAssetFilledAmount,
+            fillResults.takerAssetFilledAmount,
+            fillResults.makerFeePaid,
+            fillResults.takerFeePaid,
+            orderHash,
+            order.makerAssetData,
+            order.takerAssetData
+        );
     }
 
-    /// @dev After calling, the order can not be filled anymore.
-    /// @param order Order struct containing order specifications.
-    /// @return True if the order state changed to cancelled.
-    ///         False if the transaction was already cancelled or expired.
-    function cancelOrder(Order memory order)
-        public
-        returns (bool)
+    /// @dev Validates context for cancelOrder. Succeeds or throws.
+    /// @param order that was cancelled.
+    /// @param orderStatus Status of order that was cancelled.
+    /// @param orderHash Hash of order that was cancelled.
+    function assertValidCancel(
+        Order memory order,
+        uint8 orderStatus,
+        bytes32 orderHash
```

Comment:
assertValidFill takes in the taker address, thoughts on passing it via parameters for consistency in `assertValidCancel`?

Reply:
That would be cleaner. It would also make the function pure. `assertValidFill` is also pure, except for calling `isValidSignature`.